zkevm-circuits/src/circuit_tools/constraint_builder.rs

```diff
@@ -1,57 +1,694 @@
+    pub static_lookups: HashMap<String, Vec<LookupData<F>>>,
+    pub stored_expressions: Vec<StoredExpression<F, C>>
+}
+
+impl<F, C: CellTypeTrait> StateContext<F, C> {
+    pub fn new() -> Self {
+        StateContext {
+            constraints: Vec::new(),
+            conditions: Vec::new(),
+            dynamic_lookups: HashMap::new(),
+            dynamic_tables: HashMap::new(),
+            static_lookups: HashMap::new(),
+            stored_expressions: Vec::new(),
+        }
+    }
+}
+
+/// Constraint builder
+#[derive(Clone)]
+pub struct ConstraintBuilder<F, C: CellTypeTrait> {
+    /// Max degree of constraints
+    max_degree: usize,
+    /// Disable macro-generated description for constraints & lookups
+    /// for graph display
+    pub disable_description: bool,
+    /// Constraints to be returned to meta
+    constraints: Vec<(&'static str, Expression<F>)>,
+    /// conditions for constraints
+    stored_conditions: Vec<Expression<F>>,
+    /// The lookups generated during synthesis
+    /// assembles runtime access to RAM
+    pub dynamic_lookups: HashMap<String, Vec<LookupData<F>>>,
+    /// The tables written during synthesis
+    /// write to RAM
+    pub dynamic_tables: HashMap<String, Vec<LookupData<F>>>,
+    /// All stored expressions
+    pub static_lookups: HashMap<String, Vec<LookupData<F>>>,
+
+    pub stored_expression:  HashMap<usize, Vec<StoredExpression<F, C>>>,
+    /// CellManager
+    pub cell_manager: Option<CellManager_<F, C>>,
+    /// lookup input challenge
+    pub static_challenge: Expression<F>,
+    /// state contect
+    pub context: StateContext<F, C>,
+
+}
+
+impl<F: Field, C: CellTypeTrait> ConstraintBuilder<F, C> {
+    pub(crate) fn new(
+        max_degree: usize,
+        static_challenge: Expression<F>, 
+        cell_manager: Option<CellManager_<F, C>>
+    ) -> Self {
         ConstraintBuilder {
+            max_degree,
+            disable_description: false,
+            constraints: Vec::new(),
+            stored_conditions: Vec::new(),
+            dynamic_lookups: HashMap::new(),
+            dynamic_tables: HashMap::new(),
+            static_lookups: HashMap::new(),
+            stored_expression: HashMap::new(),
+            cell_manager,
+            static_challenge,
+            context: StateContext::new(),
+        }
+    }
+
+    /// Leave the current context by draining data to storage
+    /// save the conditions and clear the context
+    pub(crate) fn push_state(&mut self, state: usize) {
+        self.stored_conditions = self.context.conditions.clone();
+        self.constraints.append(&mut self.context.constraints);
+        self.dynamic_lookups.extend(self.context.dynamic_lookups.drain());
+        self.dynamic_tables.extend(self.context.dynamic_tables.drain());
+        self.static_lookups.extend(self.context.static_lookups.drain());
+        self.stored_expression.insert(state, self.context.stored_expressions.clone());
```

Comment:
This one, and maybe others, look a bit strange to me. To do this in `push_state` seems strange because the stored expressions still need to be added for this state, doing this in `pop_state` makes more sense to me.